transducers/shared/src/main/scala/fs2/data/stt/STT.scala

```diff
@@ -0,0 +1,263 @@
+                case (Expr1(xe), Expr0(ye)) =>
+                  loop(rest, env.update(x, Expr0.Subst(xe, ye)).update(y, Expr0.Empty()))
+                case (Expr0(xe), ye) =>
+                  F.raiseError(STTException(show"cannot substitute in an expression of type 0: $xe"))
+              }
+            case Assignment.SubstInY(x, y) =>
+              (env.lookupExpr[F](x), env.lookupExpr[F](y)).tupled.flatMap {
+                case (Expr1(xe), Expr1(ye)) =>
+                  loop(rest, env.update(x, Expr1.Subst(ye, xe)).update(y, Expr1.Hole()))
+                case (Expr0(xe), Expr1(ye)) =>
+                  loop(rest, env.update(x, Expr0.Subst(ye, xe)).update(y, Expr0.Empty()))
+                case (xe, Expr0(ye)) =>
+                  F.raiseError(STTException(show"cannot substitute in an expression of type 0: $ye"))
+              }
+            case Assignment.Swap(x, y) =>
+              (env.lookupExpr[F](x), env.lookupExpr[F](y)).tupled.flatMap { case (xe, ye) =>
+                loop(rest, env.update(x, ye).update(y, xe))
+              }
+          }
+        case Nil => F.pure(env.destackify)
+      }
+    loop(assignments, env)
+  }
+
+  private def step(q: Int, stack: List[(StackElem, Env[Variable.Stack, Out])], env: Env[Variable.Normal, Out])(
+      in: In): OptionT[F, (Int, List[(StackElem, Env[Variable.Stack, Out])], Env[Variable.Normal, Out])] =
+    in match {
+      case Tag.Internal() =>
+        OptionT.fromOption(internalTransitions.get(q -> in)).semiflatMap { case InternalTransition(q1, upd) =>
+          update(env.widen, upd).map((q1, stack, _))
+        }
+      case Tag.Call() =>
+        OptionT.fromOption(callTransitions.get(q -> in)).semiflatMap { case CallTransition(q1, p, upd) =>
+          update(env.widen, upd).map(env => (q1, (p, env.stackify) :: stack, env.call))
+        }
+      case Tag.Return() =>
+        stack match {
+          case (p, env1) :: stack =>
+            OptionT.fromOption(returnTransitions.get((q, p, in))).semiflatMap { case ReturnTransition(q1, upd) =>
+              update(env.merge(env1), upd).map((q1, stack, _))
+            }
+          case Nil =>
+            OptionT.liftF(
+              F.raiseError(STTException("inconsistent stack state. Input is probably not a well-formed tree")))
+        }
+    }
+
+  def eval0(env: Env[Variable.Normal, Out], e: Expr0[Out]): F[Chain[Out]] = {
+    def loop(e: Expr0[Out], acc: Chain[Out]): F[Chain[Out]] = {
+      e match {
+        case Expr0.Empty() =>
+          F.pure(acc)
+        case Expr0.Var(x) =>
+          env.lookup[F, Expr0[Out]](x).flatMap(loop(_, acc))
+        case Expr0.Char(c) =>
+          F.pure(acc.append(c))
+        case Expr0.Subtree(open, sub, close) =>
+          loop(sub, acc.append(open)).map(_.append(close))
+        case Expr0.Concat(left, right) =>
+          for {
+            acc <- loop(left, acc)
+            acc <- loop(right, acc)
+          } yield acc
```

Comment:
```suggestion
          loop(left, acc).flatMap(loop(right, _))
```

I find the for-comprehension with re-use of `acc` unnecessarily hard to read and it has the tail `map` problem on Scala 3 IIRC.